src/Facebook/Authentication/OAuth2Client.php

```diff
@@ -0,0 +1,298 @@
+      $this->app,
+      $this->app->getAccessToken(),
+      'GET',
+      '/debug_token',
+      $params,
+      null,
+      $this->graphVersion
+    );
+    $response = $this->client->sendRequest($this->lastRequest);
+    $metadata = $response->getDecodedBody();
+
+    return new AccessTokenMetadata($metadata);
+  }
+
+  /**
+   * Generates an authorization URL to begin the process of authenticating a user.
+   *
+   * @param string $redirectUrl The callback URL to redirect to.
+   * @param array $scope An array of permissions to request.
+   * @param string $state The CSPRNG-generated CSRF value.
+   * @param array $params An array of parameters to generate URL.
+   * @param string $separator The separator to use in http_build_query().
+   *
+   * @return string
+   */
+  public function getAuthorizationUrl($redirectUrl, array $scope = [], $state, array $params = [], $separator = '&')
+  {
+    $params += [
+      'client_id' => $this->app->getId(),
+      'state' => $state,
+      'response_type' => 'code',
+      'sdk' => 'php-sdk-' . Facebook::VERSION,
+      'redirect_uri' => $redirectUrl,
+      'scope' => implode(',', $scope)
+    ];
+
+    return static::BASE_AUTHORIZATION_URL . '/' . $this->graphVersion . '/dialog/oauth?' .
+            http_build_query($params, null, $separator);
+  }
+
+  /**
+   * Get a valid access token from a code.
+   *
+   * @param string $code
+   * @param string $redirectUri
+   * @param string|null $machineId
+   *
+   * @return AccessToken
+   *
+   * @throws FacebookSDKException
+   */
+  public function getAccessTokenFromCode($code, $redirectUri = '', $machineId = null)
+  {
+    $params = [
+      'code' => $code,
+      'redirect_uri' => $redirectUri,
+    ];
+
+    if ($machineId) {
+      $params['machine_id'] = $machineId;
+    }
+
+    return $this->requestAnAccessToken($params);
+  }
+
+  /**
+   * Exchanges a short-lived access token with a long-lived access token.
+   *
+   * @param AccessToken|string $accessToken
+   *
+   * @return AccessToken
+   *
+   * @throws FacebookSDKException
+   */
+  public function getLongLivedAccessToken($accessToken)
+  {
+    $params = [
+      'grant_type' => 'fb_exchange_token',
+      'fb_exchange_token' => (string) $accessToken,
+    ];
+
+    return $this->requestAnAccessToken($params);
+  }
+
+  /**
+   * Get a valid code from an access token.
+   *
+   * @param AccessToken|string $accessToken
+   * @param string $redirectUri
+   *
+   * @return AccessToken
+   *
+   * @throws FacebookSDKException
+   */
+  public function getCodeFromLongLivedAccessToken($accessToken, $redirectUri = '')
+  {
+    $params = [
+      'redirect_uri' => $redirectUri,
+    ];
+
+    $response = $this->sendRequestWithClientParams('/oauth/client_code', $params, $accessToken);
+    $data = $response->getDecodedBody();
+
+    if (isset($data['code'])) {
+      return $data['code'];
+    }
+
+    throw new FacebookSDKException('Code was not returned from Graph.', 401);
```

Comment:
I'd invert here too. My philosophy is: problems and early returns first, function purpose last :)